File: sdk/javascript/src/federation.js
/**
 * FoodBlock Federation — multi-server discovery, replication, and peer handshake.
 *
 * Servers publish a /.well-known/foodblock endpoint describing their capabilities.
 * Blocks can be resolved, pushed, and pulled across multiple servers.
 */

/**
 * Discover a FoodBlock server's capabilities.
 * @param {string} serverUrl - Base URL of the server
 * @param {object} [opts] - { fetch: custom fetch, timeout: ms }
 * @returns {object} Server discovery info (including public_key, signature)
 */
async function discover(serverUrl, opts = {}) {
  const fetchFn = opts.fetch || globalThis.fetch
  const timeout = opts.timeout || 10000

  const url = `${serverUrl.replace(/\/$/, '')}/.well-known/foodblock`
  const res = await fetchFn(url, {
    signal: AbortSignal.timeout(timeout)
  })

  if (!res.ok) {
    throw new Error(`FoodBlock: discovery failed for ${serverUrl}: ${res.status}`)
  }

  return res.json()
}

/**
 * Create a federated resolver that tries multiple servers.
 * Tries local first, then peers in order.
 *
 * @param {string[]} servers - Array of server URLs to try, in priority order
 * @param {object} [opts] - { fetch: custom fetch, timeout: ms, cache: boolean }
 * @returns {function} async (hash) => block | null
 */
function federatedResolver(servers, opts = {}) {
  const fetchFn = opts.fetch || globalThis.fetch
  const timeout = opts.timeout || 10000
  const cache = opts.cache !== false ? new Map() : null

  return async function resolve(hash) {
    // Check cache first
    if (cache && cache.has(hash)) {
      return cache.get(hash)
    }

    for (const server of servers) {
      try {
        const url = `${server.replace(/\/$/, '')}/blocks/${hash}`
        const res = await fetchFn(url, {
          signal: AbortSignal.timeout(timeout)
        })

        if (res.ok) {
          const block = await res.json()
          if (block && !block.error) {
            if (cache) cache.set(hash, block)
            return block
          }
        }
      } catch {
        // Try next server
        continue
      }
    }

    return null
  }
}

/**
 * Generate the well-known discovery document for a server.
 * @param {object} info - Server info
 * @returns {object} Discovery document
 */
function wellKnown(info) {
  return {
    protocol: 'foodblock',
    version: info.version || '0.4.0',
    name: info.name || 'FoodBlock Server',
    public_key: info.public_key || null,
    types: info.types || [],
    count: info.count || 0,
    schemas: info.schemas || [],
    templates: info.templates || [],
    peers: info.peers || [],
    endpoints: {
      blocks: '/blocks',
      batch: '/blocks/batch',
      chain: '/chain',
      heads: '/heads',
      push: '/.well-known/foodblock/push',
      pull: '/.well-known/foodblock/pull',
      handshake: '/.well-known/foodblock/handshake'
    }
  }
}

/**
 * Perform a peer handshake with a remote FoodBlock server.
 * Sends our identity (signed) and registers as a peer.
 *
 * @param {string} remoteUrl - Base URL of the remote server
 * @param {object} identity - { peer_url, peer_name, public_key }
 * @param {function} signFn - (payload) => signature hex string
 * @param {object} [opts] - { fetch, timeout }
 * @returns {object} Remote server's identity response
 */
async function handshake(remoteUrl, identity, signFn, opts = {}) {
  const fetchFn = opts.fetch || globalThis.fetch
  const timeout = opts.timeout || 15000

  const payload = {
    peer_url: identity.peer_url,
    peer_name: identity.peer_name,
    public_key: identity.public_key,
    timestamp: new Date().toISOString()
  }

  const signature = signFn(payload)

  const url = `${remoteUrl.replace(/\/$/, '')}/.well-known/foodblock/handshake`
  const res = await fetchFn(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      ...identity,
      payload,
      signature
    }),
    signal: AbortSignal.timeout(timeout)
  })

  if (!res.ok) {
    const body = await res.text().catch(() => '')
    throw new Error(`FoodBlock: handshake failed with ${remoteUrl}: ${res.status} ${body}`)
  }

  return res.json()
}

/**
 * Push blocks to a remote FoodBlock server.
 *
 * @param {string} remoteUrl - Base URL of the remote server
 * @param {object[]} blocks - Array of blocks to push
 * @param {object} [identity] - { peer_url, public_key }
 * @param {function} [signFn] - (payload) => signature hex string
 * @param {object} [opts] - { fetch, timeout }
 * @returns {object} { inserted, skipped, failed }
 */
async function push(remoteUrl, blocks, identity, signFn, opts = {}) {
  const fetchFn = opts.fetch || globalThis.fetch
  const timeout = opts.timeout || 30000

  const body = { blocks }

  // Sign the push if identity provided
  if (identity && signFn) {
    const payload = {
      peer_url: identity.peer_url,
      block_count: blocks.length,
      block_hashes: blocks.map(b => b.hash).filter(Boolean)
    }
    body.peer_url = identity.peer_url
    body.public_key = identity.public_key
    body.signature = signFn(payload)
  }

  const url = `${remoteUrl.replace(/\/$/, '')}/.well-known/foodblock/push`
  const res = await fetchFn(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
    signal: AbortSignal.timeout(timeout)
  })

  if (!res.ok) {
    const text = await res.text().catch(() => '')
    throw new Error(`FoodBlock: push to ${remoteUrl} failed: ${res.status} ${text}`)
  }

  return res.json()
}

/**
 * Pull blocks from a remote FoodBlock server since a cursor.
 *
 * @param {string} remoteUrl - Base URL of the remote server
 * @param {object} [query] - { since?: ISO string, after_hash?: string, types?: string[], limit?: number }
 * @param {object} [opts] - { fetch, timeout }
 * @returns {object} { blocks, count, cursor, has_more }
 */
async function pull(remoteUrl, query = {}, opts = {}) {
  const fetchFn = opts.fetch || globalThis.fetch
  const timeout = opts.timeout || 30000

  const url = `${remoteUrl.replace(/\/$/, '')}/.well-known/foodblock/pull`
  const res = await fetchFn(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(query),
    signal: AbortSignal.timeout(timeout)
  })

  if (!res.ok) {
    const text = await res.text().catch(() => '')
    throw new Error(`FoodBlock: pull from ${remoteUrl} failed: ${res.status} ${text}`)
  }

  return res.json()
}

/**
 * Bidirectional sync between two servers.
 * Pulls from remote, then pushes local blocks.
 *
 * @param {string} remoteUrl - Remote server URL
 * @param {object} localStore - { pull: fn, push: fn } or store adapter
 * @param {object} [opts] - { since, types, identity, signFn, fetch, timeout }
 * @returns {object} { pulled: { blocks, count }, pushed: { inserted, skipped, failed } }
 */
async function sync(remoteUrl, localStore, opts = {}) {
  // Pull from remote
  const pulled = await pull(remoteUrl, {
    since: opts.since,
    types: opts.types,
    limit: opts.limit || 1000
  }, opts)

  // Insert pulled blocks locally
  let localInserted = 0
  if (pulled.blocks && pulled.blocks.length && localStore.insertBlock) {
    for (const block of pulled.blocks) {
      try {
        await localStore.insertBlock(block)
        localInserted++
      } catch {
        // Skip duplicates or invalid blocks
      }
    }
  }

  // Push local blocks to remote
  let pushResult = { inserted: 0, skipped: 0, failed: 0 }
  if (localStore.getBlocksSince) {
    const localBlocks = await localStore.getBlocksSince(opts.since, opts.types)
    if (localBlocks.length) {
      pushResult = await push(remoteUrl, localBlocks, opts.identity, opts.signFn, opts)
    }
  }

  return {
    pulled: { count: pulled.count, local_inserted: localInserted, cursor: pulled.cursor },
    pushed: pushResult
  }
}

module.exports = { discover, federatedResolver, wellKnown, handshake, push, pull, sync }

File: sdk/swift/Sources/Federation.swift
import Foundation

/// FoodBlock Federation — multi-server discovery, replication, and peer handshake.
///
/// Servers publish a `/.well-known/foodblock` endpoint describing their capabilities.
/// Blocks can be resolved, pushed, and pulled across multiple servers.

public enum FoodBlockFederation {

    // MARK: - Types

    /// Standard endpoint paths for a FoodBlock server.
    public struct WellKnownEndpoints: Codable, Equatable {
        public let blocks: String
        public let batch: String
        public let chain: String
        public let heads: String
        public let push: String
        public let pull: String
        public let handshake: String

        public init(
            blocks: String = "/blocks",
            batch: String = "/blocks/batch",
            chain: String = "/chain",
            heads: String = "/heads",
            push: String = "/.well-known/foodblock/push",
            pull: String = "/.well-known/foodblock/pull",
            handshake: String = "/.well-known/foodblock/handshake"
        ) {
            self.blocks = blocks
            self.batch = batch
            self.chain = chain
            self.heads = heads
            self.push = push
            self.pull = pull
            self.handshake = handshake
        }
    }

    /// The full discovery document published at `/.well-known/foodblock`.
    public struct WellKnownDoc: Codable, Equatable {
        public let `protocol`: String
        public let version: String
        public let name: String
        public let types: [String]
        public let count: Int
        public let schemas: [String]
        public let templates: [String]
        public let peers: [String]
        public let endpoints: WellKnownEndpoints

        public init(
            protocol: String = "foodblock",
            version: String = "0.4.0",
            name: String = "FoodBlock Server",
            types: [String] = [],
            count: Int = 0,
            schemas: [String] = [],
            templates: [String] = [],
            peers: [String] = [],
            endpoints: WellKnownEndpoints = WellKnownEndpoints()
        ) {
            self.`protocol` = `protocol`
            self.version = version
            self.name = name
            self.types = types
            self.count = count
            self.schemas = schemas
            self.templates = templates
            self.peers = peers
            self.endpoints = endpoints
        }
    }

    // MARK: - Well-Known Document

    /// Generate the well-known discovery document for a server.
    ///
    /// - Parameters:
    ///   - name: Server name. Defaults to "FoodBlock Server".
    ///   - version: Protocol version. Defaults to "0.4.0".
    ///   - types: Block types this server supports.
    ///   - count: Total number of blocks stored.
    ///   - schemas: Schema hashes this server recognizes.
    ///   - templates: Template hashes available.
    ///   - peers: URLs of known peer servers.
    /// - Returns: A `WellKnownDoc` ready for JSON serialization.
    public static func wellKnown(
        name: String = "FoodBlock Server",
        version: String = "0.4.0",
        types: [String] = [],
        count: Int = 0,
        schemas: [String] = [],
        templates: [String] = [],
        peers: [String] = []
    ) -> WellKnownDoc {
        return WellKnownDoc(
            protocol: "foodblock",
            version: version,
            name: name,
            types: types,
            count: count,
            schemas: schemas,
            templates: templates,
            peers: peers,
            endpoints: WellKnownEndpoints()
        )
    }

    // MARK: - Discovery

    /// Discover a FoodBlock server's capabilities by fetching its well-known endpoint.
    ///
    /// - Parameter serverUrl: Base URL of the server (e.g., "https://api.example.com").
    /// - Throws: An error if the request fails or the response cannot be decoded.
    /// - Returns: The server's `WellKnownDoc` describing its capabilities, types, and peers.
    public static func discover(serverUrl: String) async throws -> WellKnownDoc {
        let trimmed = serverUrl.hasSuffix("/")
            ? String(serverUrl.dropLast())
            : serverUrl

        guard let url = URL(string: "\(trimmed)/.well-known/foodblock") else {
            throw FederationError.invalidURL(serverUrl)
        }

        var request = URLRequest(url: url)
        request.timeoutInterval = 10

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw FederationError.discoveryFailed(serverUrl, 0)
        }

        guard httpResponse.statusCode >= 200, httpResponse.statusCode < 300 else {
            throw FederationError.discoveryFailed(serverUrl, httpResponse.statusCode)
        }

        let decoder = JSONDecoder()
        return try decoder.decode(WellKnownDoc.self, from: data)
    }

    // MARK: - Federated Resolver

    /// Create a federated resolver that tries multiple servers in priority order.
    ///
    /// Returns a closure compatible with `FoodBlockChain.Resolver` that attempts to
    /// fetch a block from each server in sequence, returning the first successful result.
    ///
    /// - Parameter servers: Array of server base URLs, in priority order (local first).
    /// - Returns: An async closure `(String) async -> FoodBlock?` that resolves block hashes.
    public static func federatedResolver(
        servers: [String]
    ) -> (String) async -> FoodBlock? {
        return { hash in
            for server in servers {
                let trimmed = server.hasSuffix("/")
                    ? String(server.dropLast())
                    : server

                guard let url = URL(string: "\(trimmed)/blocks/\(hash)") else {
                    continue
                }

                var request = URLRequest(url: url)
                request.timeoutInterval = 10

                do {
                    let (data, response) = try await URLSession.shared.data(for: request)

                    guard let httpResponse = response as? HTTPURLResponse,
                          httpResponse.statusCode >= 200,
                          httpResponse.statusCode < 300 else {
                        continue
                    }

                    let decoder = JSONDecoder()
                    let block = try decoder.decode(FoodBlock.self, from: data)
                    return block
                } catch {
                    continue
                }
            }
            return nil
        }
    }

    // MARK: - Errors

    public enum FederationError: Error, CustomStringConvertible {
        case invalidURL(String)
        case discoveryFailed(String, Int)

        public var description: String {
            switch self {
            case .invalidURL(let url):
                return "FoodBlock: invalid server URL '\(url)'"
            case .discoveryFailed(let url, let status):
                return "FoodBlock: discovery failed for \(url): \(status)"
            }
        }
    }
}
